File: 1046-max-consecutive-ones-iii/max-consecutive-ones-iii.java
class Solution {
    public int longestOnes(int[] nums, int k) {
        int longest = 0;
        int curr = 0;

        int flipCnt = 0;
        int start = 0;
        for (int i = 0; i < nums.length; i++) {
            if (nums[i] == 0 && flipCnt == k) {
                longest = Math.max(longest, curr);
                while (flipCnt == k) {
                    if (nums[start] == 0)
                        flipCnt--;
                    curr--;
                    start++;
                }
            }
            if (nums[i] == 0)
                flipCnt++;
            curr++;
        }

        return Math.max(longest, curr);
    }
}

File: 1046-max-consecutive-ones-iii/max-consecutive-ones-iii.js
/**
 * @param {number[]} nums
 * @param {number} k
 * @return {number}
 */
var longestOnes = function (nums, k) {
    let left = 0, maxConsecutive = 0, zeroCount = 0;

    for (let right = 0; right < nums.length; right++) {
        if (nums[right] === 0) zeroCount++;

        while (zeroCount > k) {
            if (nums[left] === 0) zeroCount--;
            left++;
        }

        maxConsecutive = Math.max(maxConsecutive, right - left + 1);
    }

    return maxConsecutive;
};
